readme.md
[![Jikan](http://i.imgur.com/ctoJ3Jp.png)](#jikan---unofficial-myanimelistnet-php-api)

# Jikan - Unofficial MyAnimeList.net PHP API

[![build](https://travis-ci.org/jikan-me/jikan.svg?branch=master)](https://travis-ci.org/jikan-me/jikan?branch=master) [![build](https://ci.appveyor.com/api/projects/status/github/jikan-me/jikan?branch=master&svg=true)](https://ci.appveyor.com/project/irfan-dahir/jikan) [![stable](https://img.shields.io/packagist/v/jikan-me/jikan.svg?style=flat)](https://packagist.org/packages/jikan-me/jikan) [![Average time to resolve an issue](http://isitmaintained.com/badge/resolution/jikan-me/jikan.svg)](http://isitmaintained.com/project/jikan-me/jikan "Average time to resolve an issue") [![Percentage of issues still open](http://isitmaintained.com/badge/open/jikan-me/jikan.svg)](http://isitmaintained.com/project/jikan-me/jikan "Percentage of issues still open") [![stable](https://img.shields.io/badge/PHP-^%207.1-blue.svg?style=flat)]() [![Discord Server](https://img.shields.io/discord/460491088004907029.svg?style=flat&logo=discord)](https://discord.gg/4tvCr36)

Jikan is a PHP API for [MyAnimeList.net](https://myanimelist.net). It scrapes the website to satisfy the need for an API - which MyAnimeList lacks.

The raison d'être of Jikan is to assist developers easily get the data they need for their apps and projects without having to depend on the lackluster official API, unstable APIs, or sidetracking their projects to develop parsers.

The word _Jikan_ literally translates to _Time_ in Japanese (**時間**). And that's what this API saves you of. ;)

**Notice**: Jikan does not support authenticated requests. You can not update your lists.

## Getting Started

| Version   | Support | PHP | Lumen/Laravel |
|------------|----------|----------|----------| 
| [`^3` (in Beta)](https://github.com/jikan-me/jikan/tree/3.0.0) | ✅ New features | [![beta](https://img.shields.io/badge/PHP-^%207.4-blue.svg?style=flat)]() | `^8` |
| [`^2` (master)](https://github.com/jikan-me/jikan)      | ⚠️ Maintenance only | [![stable](https://img.shields.io/badge/PHP-^%207.1-blue.svg?style=flat)]() | `^6`
| [`~1`](https://github.com/jikan-me/jikan/tree/1.16.3)      | ❌ No longer maintained or supported | [![stable](https://img.shields.io/badge/PHP-^%207.0-blue.svg?style=flat)]() | `5.5.*` |


1. `composer require jikan-me/jikan` (This will install  version `^2`)
2. [Documentation](http://docs.jikan.moe)

# Jikan REST API

A REST API service is available as well

- **[REST API DOCUMENTATION](https://jikan.docs.apiary.io)**
- **[Apps/Projects using the REST API](https://jikan.moe/showcase)**
- **[Host the REST API yourself](https://github.com/jikan-me/jikan-rest)**

## Wrappers

| Language   | Wrappers |
|------------|----------|
| JavaScript | [JikanJS](https://github.com/zuritor/jikanjs) by Zuritor |
| Java       | [Jikan4java](https://github.com/Doomsdayrs/Jikan4java) by Doomsdayrs<br>[reactive-jikan](https://github.com/SandroHc/reactive-jikan) by Sandro Marques<br>[Jaikan](https://github.com/ShindouMihou/Jaikan) by ShindouMihou |
| Python     | [JikanPy](https://github.com/abhinavk99/jikanpy) by Abhinav Kasamsetty |
| Node.js    | [jikan-node](https://github.com/xy137/jikan-node) by xy137<br>[jikan-nodejs](https://github.com/ribeirogab/jikan-nodejs) by ribeirogab |
| TypeScript | [jikants](https://github.com/Julien-Broyard/jikants) by Julien Broyard<br>[jikan-client](https://github.com/javi11/jikan-client) by Javier Blanco |
| PHP        | [jikan-php](https://github.com/janvernieuwe/jikan-jikanPHP) by Jan Vernieuwe |
| .NET       | [Jikan.net](https://github.com/Ervie/jikan.net) by Ervie |
| Elixir     | [JikanEx](https://github.com/seanbreckenridge/jikan_ex) by Sean Breckenridge |
| Go         | [jikan-go](https://github.com/darenliang/jikan-go) by Daren Liang<br>[jikan2go](https://github.com/nokusukun/jikan2go) by nokusukun |
| Ruby       | [Jikan.rb](https://github.com/Zerocchi/jikan.rb) by Zerocchi |
| Dart       | [jikan-dart](https://github.com/charafau/jikan-dart) by Rafal Wachol |
| Kotlin     | [JikanKt](https://github.com/GSculerlor/JikanKt) by Ganedra Afrasya |

[Add your wrapper here](https://github.com/jikan-me/jikan/edit/master/readme.md)

# Features

- Anime
  - Main Information
  - Characters & Staff
  - Episodes
  - Episode Details
  - News
  - Videos/PV/Episodes
  - Pictures
  - Stats
  - Forum Topics
  - More Info
  - Recommendations
  - Reviews
  - Recent List Updates By Users
- Manga
  - Main Information
  - Characters
  - News
  - Stats
  - Pictures
  - Forum Topics
  - More Info
  - Recommendations
  - Reviews
  - Recent List Updates By Users
- Character
  - Main Information
  - Pictures
- People
  - Main Information
  - Pictures
- Search
  - Anime
  - Manga
  - Character
  - Person
  - Pagination Support
  - Advanced Search
    - Filters
    - Order By
    - Sorting (Ascending/Descending)
- Seasonal Anime (Season + Year)
- Season List/Archive
- Anime Scheduling (for current season)
- Top
  - Anime
  - Manga
  - Characters
  - People
  - Sub Types & Pagination Support
- Genre
  - Anime Listing (All Anime by Genre)
  - Anime Genre Listing (All Genres + metadata)
  - Manga Listing (All Anime by Genre)
  - Manga Genre Listing (All Genres + metadata)
- Producer
  - Anime Listing (All Anime by a Producer)
  - Producers Listing (All Producers + metadata)
  - Manga Listing (All Manga by a Producer)
  - Magazines Listing (All Magazines + metadata)
- User
  - Profile
  - Friends
    - Pagination support
  - History
    - All
    - Anime
    - Manga
  - Anime & Manga Lists
    - Pagination Support
- Club
  - Main Information
  - User List

[View RoadMap](https://github.com/jikan-me/jikan/projects/4)

# Running Tests

## PHPUnit

`php vendor/bin/phpunit`

## GrumPHP

PHPCS, PHPLint & PHPUnit

`php vendor/bin/grumphp run`

---

# Backers

A huge thank you to all our Patrons! 🙏 This project wouldn't be running without your support.

We have a free [REST API service](https://jikan.moe), if you wish to support us you can [become a Patron!](https://patreon.com/jikan)

## Sugoi (すごい) Patrons

- [Jared Allard (jaredallard)](https://github.com/jaredallard)
- [hugonun (hug_onun)](https://twitter.com/hug_onun)

## Patrons

- Aaron Treinish
- Cesar Irad Mendoza
- Jason Weatherly
- Jesse
- Kundan Chintamaneni
- Kururin
- Purplepinapples
- Sakamotodesu
- TeraNovaLP

## Development

|||
|------------|----------|
| ![JetBrain](https://user-images.githubusercontent.com/9166451/126047249-9e5bdc63-ae91-4082-bca5-ffe271b421da.png) | Jikan's development is powered by [JetBrain's Open Source License](https://jb.gg/OpenSource) |

A shoutout to their amazing products and for supporting Jikan since early versions!

---

# Release Changelog

## 2.17.0 - Sep 16, 21

- Added support for MAL's new [genre overhaul](https://myanimelist.net/forum/?topicid=1956762): `themes`, `demographics`, `explicitGenres`
- Updated Constants to reflect new and modified genres (while retaining BC). Some genres have been deleted or merged with others so they may return 404, check [Constants](https://github.com/jikan-me/jikan/blob/62f3e12cbcc8d841b3f923e4317f0b50f28f0574/src/Helper/Constants.php) for details
- Anime/Manga Genres List now returns additional arrays for `themes`, `demographics`, `explicitGenres` as they have been split up from `genres`
- Added parser support for Anime and Manga External Links https://github.com/jikan-me/jikan/issues/353
- Parser bug fixes

[Read More](https://github.com/jikan-me/jikan/blob/master/changelog.md)

# DISCLAIMER

- Jikan is not affiliated with MyAnimeList.net
- You are responsible for the usage of this API. Please be respectful towards MyAnimeList's [Terms Of Service](https://myanimelist.net/about/terms_of_use)


changelog.md
## Changelog

### 

## 2.17.0 - Sep 16, 21

- Added support for MAL's new [genre overhaul](https://myanimelist.net/forum/?topicid=1956762): `themes`, `demographics`, `explicitGenres`
- Updated Constants to reflect new and modified genres (while retaining BC). Some genres have been deleted or merged with others so they may return 404, check [Constants](https://github.com/jikan-me/jikan/blob/62f3e12cbcc8d841b3f923e4317f0b50f28f0574/src/Helper/Constants.php) for details
- Anime/Manga Genres List now returns additional arrays for `themes`, `demographics`, `explicitGenres` as they have been split up from `genres`
- Added parser support for Anime and Manga External Links https://github.com/jikan-me/jikan/issues/353
- Parser bug fixes

### 2.16.0 - Jan 27, 20
- Added **Top Reviews**(Anime, Manga & Best Voted (includes both)) parsing [#251](/../../issues/251)
```php

/**
 * Type Constants
 * Anime: \Jikan\Helper\Constants::TOP_REVIEW_ANIME
 * Manga: \Jikan\Helper\Constants::TOP_REVIEW_MANGA
 * Best Voted (Anime + Manga): \Jikan\Helper\Constants::TOP_REVIEW_BEST_VOTED 
 */
$type = \Jikan\Helper\Constants::TOP_REVIEW_ANIME;
$page = 1;

$topReviews = $jikan->getTopReviews(
    new \Jikan\Request\Top\TopReviewsRequest($type, $page)
);
```

- Bug fixes [#249](/../../issues/249), [#250](/../../issues/250), [#259](/../../issues/259)

### 2.15.0 - Jan 10, 20
- Add **Producer List** parsing
```php
$producers = $jikan->getProducers(
    new \Jikan\Request\Producer\ProducersRequest()
);
```
- Add **Magazine List** parsing
```php
$magazines = $jikan->getMagazines(
    new \Jikan\Request\Magazine\MagazinesRequest()
);
```
- Add **Anime Genre List** parsing
```php
$genres = $jikan->getAnimeGenres(
    new \Jikan\Request\Genre\AnimeGenresRequest()
);
```

- Add **Manga Genre List** parsing
```php
$genres = $jikan->getMangaGenres(
    new \Jikan\Request\Genre\MangaGenresRequest()
);
```

- Parser bug fixes
- Improved image quality parsing for some requests
  `\Jikan\Helper\Parser::parseImageQuality($imageUrl)` now removes `v` from the end of images, which resulted in a smaller thumbnail


### 2.9.0 - Nov 11, 19
- **[Anime Episode Details]**
	- Added feature request [jikan#229](https://github.com/jikan-me/jikan/issues/229)
- **[User Profile]**
	- Added `userId` property [jikan-rest#74](https://github.com/jikan-me/jikan-rest/issues/74)

### 2.8.6 - Aug 18, 19
- Bug fixes
- **[Search]**
	- Throws `BadResponseException` if search query is < 3

### 2.8.0 - Apr 30, 19
- **[User List : Anime|Manga]**
	- Throw `BadResponseException` instead of `ParserException` when the request fails [#217](/../../issues/217)
	- Add list filtering & sorting [jikan-rest#49](https://github.com/jikan-me/jikan-rest/issues/49)
	- Replace `StudioMeta`, `LicensorMeta` & `MagazineMeta` with `MalUrl` intead
		- This will further provide properties like `type`, `url`
- **[Search : Anime|Manga]**
	- Add result ordering & sorting [jikan-rest#48](https://github.com/jikan-me/jikan-rest/issues/48)

### 2.7.0 - Apr 14, 19
- **[Anime|Manga]**
	- `EpisodeListItem` `aired` property to return `DateTimeImmutable` instead of `DateRange`
	- [Bug Fix] Stats Score refactoring; returns `AnimeStatsScore`/`MangaStatsScore` object instead of array now - [#216](/../../issues/216)
	- Fix `AnimeListItem`/`MangaListItem` parsing date formats incorrectly
- **[Helper]** fix `parseDateMDY` returning incorrect date
- **[Jikan.php]**
	- fix `PersonSearch`incorrect Request class hinting
	- add `UserMangaList`
- **[MalClient.php]**
	- `getAnimeEpisodes` with no episodes return empty episodes array now instead of BadResponseException (404)
	- `getPersonSearch` with no results returns empty array now instead of BadResponseException (404)
	- `getCharacterSearch` with no results returns empty array now instead of BadResponseException (404)
	- `getCharacter` returns BadResponseException (404) now if invalid ID is provided
	- Fix guzzle dep injection being overriden in some cases
- **[User]**
	- Fix `UserProfileParser`'s `getUsername()` returning the URL instead of the username
- Added default values to some models
- Other minor code and bug fixes

### 2.6.1 stable - Dec 17, 18
- **[Club]** Optimize User list schema

### 2.6.0 stable - Dec 16, 18
- **[Club]** Add club parsing

### 2.5.0 stable - Dec 13, 18
- **[Anime|Manga]** Add **Reviews** parsing
- **[Club]** Add **User List** parsing - 2.4.0

### 2.3.0 stable - Nov 14, 18
- **[Anime|Manga]** Add **Recommendations** parsing
- **[Anime|Manga]** Add **Recently Updated By Users** parsing - 2.2.0
- **[Seasonal]** Added parsing for **later** anime - 2.1.4
- **[User List]** Bug fix for some manga lists parsing - 2.1.3

### 2.1.2 stable - Oct 20, 18
- **[Search]** Bug fix [#202](/../../issues/202) (PR[#206](/../../issues/206))

### 2.1.1 stable - Oct 12, 18
- **[Seasonal]** Parser bug fix [#205](/../../issues/205)

### 2.1.0 stable - Oct 1, 18
- **[User]** Add Anime/Manga list parsing
- Bug fixes for MalId parsing

### 2.0.0 - Sep 25, 18
- Lots of bug fixes
- Schema fixes
- All images URLs returned are now HQ
- Added DateProp for DateRange

### 2.0.0-rc.2 - Aug 30, 18
- A lot of bug fixes
- A lot of Schema improvements and fixes
- More tests coverage (fixtures added a dev dependencies)
- Reviewed syntax & usage
- Added Season List parsing
- Added more Constants

### 2.0.0-rc.1 - Aug 5, 18
- Complete refactor using better dependencies
	- Parser is more robust and stable
	- Library now implements PSR2
- Clean data values
- Added **Season List**
- Added **Top**
	- Characters
	- People
- Added **Genre**
	- Anime listing by genre
	- Manga listing by genre
- Added **Producers**
	- Anime listing by producer
- Added **Magazine**
	- Manga listing by magazine
- Added **User**
	- Profile
	- History
	- Friends
- Complete syntax revamp
- And a whole lot more!

**You'll have to visit the NEW documentation for v2+ for usage**

### 1.15.14 stable - May 26, 18
- **[Anime, Manga, Extended]**

src/consts.php
<?php

/*
 * URL & Endpoints
 */
define( 'BASE_URL' , 'https://myanimelist.net/' );

define( 'ANIME_ENDPOINT' , 'anime/' );
define( 'MANGA_ENDPOINT' , 'manga/' );
define( 'PEOPLE_ENDPOINT' , 'people/' );
define( 'CHARACTER_ENDPOINT' , 'character/' );
define( 'USER_LIST_ENDPOINT' , 'malappinfo.php');


/*
 * Request
 */
define( 'CHARACTERS_STAFF' , 'characters' );
define( 'CHARACTERS' , 'characters' );
define( 'EPISODES' , 'episode' );
define( 'NEWS' , 'news' );
define( 'VIDEOS' , 'video' );
define( 'STATS' , 'stats' );
define( 'PICTURES' , 'pics' );
define( 'FORUM' , 'forum' );
define( 'MORE_INFO' , 'moreinfo' );

define( 'ANIME' , 'Anime' );
define( 'MANGA' , 'Manga' );
define( 'CHARACTER' , 'Character' );
define( 'PERSON' , 'Person' );
define( 'PEOPLE', 'Person'); // alias

define( 'SEASON_WINTER' , 'winter' );
define( 'SEASON_SPRING' , 'spring' );
define( 'SEASON_SUMMER' , 'summer' );
define( 'SEASON_FALL' , 'fall' );

/*
 * Search
 */

// GENRE
define('GENRE_ACTION', 1);
define('GENRE_ADVENTURE', 2);
define('GENRE_CARS', 3);
define('GENRE_COMEDY', 4);
define('GENRE_DEMENTIA', 5);
define('GENRE_DEMONS', 6);
define('GENRE_MYSTERY', 7);
define('GENRE_DRAMA', 8);
define('GENRE_ECCHI', 9);
define('GENRE_FANTASY', 10);
define('GENRE_GAME', 11);
define('GENRE_HENTAI', 12);
define('GENRE_HISTORICAL', 13);
define('GENRE_HORROR', 14);
define('GENRE_KIDS', 15);
define('GENRE_MAGIC', 16);
define('GENRE_MARTIAL_ARTS', 17);
define('GENRE_MECHA', 18);
define('GENRE_MUSIC', 19);
define('GENRE_PARODY', 20);
define('GENRE_SAMURAI', 21);
define('GENRE_ROMANCE', 22);
define('GENRE_SCHOOL', 23);
define('GENRE_SCI_FI', 24);
define('GENRE_SHOUJO', 25);
define('GENRE_SHOUJO_AI', 26);
define('GENRE_SHOUNEN', 27);
define('GENRE_SHOUNEN_AI', 28);
define('GENRE_SPACE', 29);
define('GENRE_SPORTS', 30);
define('GENRE_SUPER_POWER', 31);
define('GENRE_VAMPIRE', 32);
define('GENRE_YAOI', 33);
define('GENRE_YURI', 34);
define('GENRE_HAREM', 35);
define('GENRE_SLICE_OF_LIFE', 36);
define('GENRE_SUPERNATURAL', 37);
define('GENRE_MILITARY', 38);
define('GENRE_POLICE', 39);
define('GENRE_PSYCHOLOGICAL', 40);
define('GENRE_THRILLER', 41);
define('GENRE_SEINEN', 42);
define('GENRE_JOSEI', 43);
define('GENRE_DOUJINSHI', 43);
define('GENRE_GENDER_BENDER', 44);

// Type
define('TYPE_TV', 1);
define('TYPE_OVA', 2);
define('TYPE_MOVIE', 3);
define('TYPE_SPECIAL', 4);
define('TYPE_ONA', 5);
define('TYPE_MUSIC', 6);

define('TYPE_MANGA', 1);
define('TYPE_NOVEL', 2);
define('TYPE_ONE_SHOT', 3);
define('TYPE_DOUJINSHI', 4);
define('TYPE_MANHWA', 5);
define('TYPE_MANHUA', 6);
define('TYPE_OEL', 7);

// Status
define('STATUS_AIRING', 1);
define('STATUS_FINISHED_AIRING', 2);
define('STATUS_TO_BE_AIRED', 3);

define('STATUS_PUBLISHING', 1);
define('STATUS_FINISHED_PUBLISHING', 2);
define('STATUS_TO_BE_PUBLISHED', 3);

// Rating
define('RATING_G', 1); //  All Ages
define('RATING_PG', 2); // Children
define('RATING_PG13', 2); // Teens 13 or older
define('RATING_R17', 4); // Violence & Profanity
define('RATING_R', 5); // Mild Nudity
define('RATING_RX', 6); // Hentai

/*
 * Top Anime/Manga
 */

// Anime
define('TOP_AIRING', 'airing');
define('TOP_UPCOMING', 'upcoming');
define('TOP_TV', 'tv');
define('TOP_MOVIE', 'movie');
define('TOP_OVA', 'ova');
define('TOP_SPECIAL', 'special');

// Manga
define('TOP_MANGA', 'manga');
define('TOP_NOVEL', 'novels');
define('TOP_ONE_SHOT', 'oneshots');
define('TOP_DOUJINSHI', 'doujin');
define('TOP_MANHWA', 'manhwa');
define('TOP_MANHUA', 'manhua');

// Both
define('TOP_POPULARITY', 'bypopularity');
define('TOP_FAVORITE', 'favorite');

/*
 * User
 */
define('USER_PROFILE', 'user_profile');
define('USER_HISTORY', 'user_history');
define('USER_FRIENDS', 'user_friends');

/*
 * User List
 */
define('USER_LIST_ALL', 'user_list_all');
define('USER_LIST_WATCHING', 'user_list_watching');
define('USER_LIST_COMPLETED', 'user_list_completed');
define('USER_LIST_ONHOLD', 'user_list_onhold');
define('USER_LIST_DROPPED', 'user_list_dropped');
define('USER_LIST_PLAN_TO_WATCH', 'user_list_plan_to_watch');
define('USER_LIST_PTW', 'user_list_plan_to_watch');